src/coldfront_plugin_api/urls.py

```diff
@@ -10,16 +12,72 @@
 
 
 class AllocationViewSet(viewsets.ReadOnlyModelViewSet):
+    """
+    This viewset implements the API to Coldfront's allocation object
+    The API allows filtering allocations by any of Coldfront's allocation model attributes,
+    as well as by Allocation Attributes.
+
+    Filtering by allocation model attributes
+    (i.e by pk, start_date, or even by attributes that span relationships, like project PI email)
+    is possible by using the query parameters directly as Django query filters.
+
+    For example, to filter by the allocation's project PI email: "/api/allocations?project__pi__email=test@bu.edu"
+
+    Filtering using this method requires knowledge of the Allocation model schema and Django's filters
+    Documentation of Coldfront Allocation model: https://coldfront.readthedocs.io/en/latest/apidocs/allocations/
+    Documentation for Django's query syntax: https://docs.djangoproject.com/en/5.0/topics/db/queries/#lookups-that-span-relationships
+
+    To filter by allocation attribute (AA) (i.e Quota Limits CPU),
+    name the query as the AA (case insensitive), prefixed with "attr_"
+
+    For all filters, you can also filter with conditional OR,
+    by repeating the same query field with different values
+
+    For example, to find allocations with AA Quota Limits CPU equal to 2 OR 4:
+    "/api/allocations?attr_quota_limits_cpu=2&attr_quota_limits_cpu=4"
+
+    In cases where an invalid model attribute or AA is queried, an empty list is returned
+    """
+
     serializer_class = serializers.AllocationSerializer
     authentication_classes = auth.AUTHENTICATION_CLASSES
     permission_classes = [IsAdminUser]
 
     def get_queryset(self):
         queryset = Allocation.objects.filter(status__name="Active")
+        query_params = self.request.query_params
 
-        if self.request.query_params.get("all") == "true":
+        if query_params.get("all") == "true":
             queryset = Allocation.objects.all()
 
+        allocation_attr_prefix = "attr_"
+
+        for query, val in query_params.items():
+            if query == "all":
+                continue
+            if query.startswith(allocation_attr_prefix):
+                attribute_query = query[len(allocation_attr_prefix) :]
+                val_list = query_params.getlist(query)
+                q_query = Q(allocationattribute__value=val_list[0])
+                for val in val_list[1:]:
```

Comment:
what's `val_list` supposed to look like? what's in `val_list[0]` and what's in the rest of it that we are iterating over? I am a little confused.

Reply:
`val_list` is the list of values given for a query parameter.
I.e if the user queried this endpoint: `/api/allocations?a=1&a=2&b=3`, note how the query parameter `a` is used twice, so the `val_list` for `a` will contain `[1,2]`, and for `b` it will be `[3]`.

Reply:
@QuanMPhm I think a better way to do this might be:

```python
q_query = Q()   # An empty Q query

for val in val_list:
    q_query = query | Q(allocationattribute__value=val)
```

I think this reads better. Definitely test this first because what I was just playing around in the interpreter.

Reply:
You're right, initializing with an empty Q query is fine for Django.